src/core/radio/trel_link.cpp

```diff
@@ -373,8 +376,42 @@ void Link::ProcessReceivedPacket(Packet &aPacket)
 
     Get<Mac::Mac>().HandleReceivedFrame(&mRxFrame, kErrorNone);
 
+    // As the received frame is processed by the MAC or MLE layers,
+    // `CheckPeerAddrOnRxSuccess()` may be called with different modes,
+    // depending on whether the frame passes receive security checks
+    // at either the MAC or MLE layers, allowing or disallowing peer
+    // socket address to be updated from received TREL packet info.
+
 exit:
-    return;
+    mRxPacketPeer = nullptr;
+}
+
+void Link::CheckPeerAddrOnRxSuccess(PeerSockAddrUpdateMode aMode)
+{
+    Ip6::SockAddr prevSockAddr;
+
+    VerifyOrExit(mState != kStateDisabled);
+
+    VerifyOrExit(mRxPacketPeer != nullptr);
+
+    prevSockAddr = mRxPacketPeer->GetSockAddr();
+    VerifyOrExit(prevSockAddr != mRxPacketSenderAddr);
+
+    LogNote("Peer %s rx sock-addr differs from previously discovered by platform",
+            mRxPacketPeer->GetExtAddress().ToString().AsCString());
+    LogNote("    Rcvd sock-addr:%s", mRxPacketSenderAddr.ToString().AsCString());
+    LogNote("    Prev sock-addr:%s", prevSockAddr.ToString().AsCString());
+
+    if (aMode == kAllowPeerSockAddrUpdate)
+    {
+        LogNote("Updating the peer sock-addr to the newly received");
+        mRxPacketPeer->SetSockAddr(mRxPacketSenderAddr);
+    }
+
+    Get<Interface>().NotifyPeerSocketAddressDifference(prevSockAddr, mRxPacketSenderAddr);
```

Comment:
In case of long-lived queries, can the wrong address be over-written from the mDNS cache? In addition to notification, we also need to stop any ongoing query and start a new one to avoid overwriting with a stale address again right?

Reply:
The `NotifyPeerSocketAddressDifference()` API notifies the platform layer when a discrepancy is detected between a TREL peer's IPv6 address (from a received packet) and the information previously reported by the platform layer (through DNS-SD discovery) for the same peer.

It is then up to the platform implementation to take any necessary action, which may depend on the mDNS library used, how it is integrated, and its available APIs.
- The platform may choose to ignore this and let it be fixed when secure messages are exchanged between the nodes on the TREL link (which would update the address) or when the DNS-SD TTL expires and triggers a new mDNS query to refresh the information.
- Ideally, the platform layer should send an mDNS query to confirm and/or re-discover up-to-date information about the peer (its service and addresses). How this is done depends on the platform and its mDNS library (e.g., it can stop/restart a previous long-lived query or use other specific available APIs).

Reply:
Also for the "flip-flopping" corner case, the platform should be smart enough not to retrigger an mDNS query each time this API function is called.

If the new address is already in the discovered list of addresses, that's fine. And if a query is already ongoing, don't start a new one again.